File: org.openhab.binding.vera/src/main/java/org/openhab/binding/vera/luup/Device.java
package org.openhab.binding.vera.luup;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import com.fasterxml.jackson.annotation.JsonAnyGetter;
import com.fasterxml.jackson.annotation.JsonAnySetter;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;

import static java.util.Collections.unmodifiableMap;

public class Device {
    private int id;

    private String name;

    @JsonProperty("altid")
    private int alternateId;

    @JsonProperty("parent")
    private int parentId;

    @JsonProperty("category")
    private int categoryId;

    @JsonProperty("subcategory")
    private int subcategoryId;

    @JsonProperty("room")
    private int roomId = -1;

    private Map<String, String> properties = new HashMap<>();

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public int getAlternateId() {
        return alternateId;
    }

    public int getParentId() {
        return parentId;
    }

    public int getCategoryId() {
        return categoryId;
    }

    public int getSubcategoryId() {
        return subcategoryId;
    }

    public int getRoomId() {
        return roomId;
    }

    @JsonIgnore
    public Set<String> getPropertyKeys() {
        return properties.keySet();
    }

    @JsonAnySetter
    private void setUnknownProperty(String key, String value) {
        properties.put(key, value);
    }

    @JsonAnyGetter
    private Map<String, String> getProperties() {
        return unmodifiableMap(properties);
    }

    public String getProperty(String name) {
        return properties.get(name);
    }
}

File: org.openhab.binding.vera/src/main/java/org/openhab/binding/vera/luup/Device.scala
package org.openhab.binding.vera.luup

import java.{util => ju}

import scala.beans.BeanProperty
import scala.collection.JavaConversions._

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty

class Device {
    @BeanProperty
    protected var id: Int = _

    @BeanProperty
    protected var name: String = _

    @BeanProperty
    @JsonProperty("altid")
    protected var alternateId: Int = _

    @BeanProperty
    @JsonProperty("parent")
    protected var parentId: Int = _

    @BeanProperty
    @JsonProperty("category")
    protected var categoryId: Int = _

    @BeanProperty
    @JsonProperty("subcategory")
    protected var subcategoryId: Int = _

    @BeanProperty
    @JsonProperty("room")
    protected var roomId: Int = -1

    protected var properties: Map[String, String] = Map.empty

    override def toString(): String = s"Device[$id]: $name"

    @JsonIgnore
    def getPropertyKeys(): ju.Set[String] = properties.keySet

    @JsonAnySetter
    protected def setUnknownProperty(key: String, value: String) = properties += (key -> value)

    @JsonAnyGetter
    def getProperties: ju.Map[String, String] = properties

    def getProperty(name: String) = properties(name)

    def hasProperty(name: String) = properties.isDefinedAt(name)

    def mergeWith(other: Device): Device = {
        if (other.name != null) this.name = other.name
        if (other.properties != null) this.properties = other.properties
        this.alternateId = other.alternateId
        this.parentId = other.parentId
        this.categoryId = other.categoryId
        this.subcategoryId = other.subcategoryId
        this.roomId = other.roomId
        this
    }
}
